File: CrackTheCodingInterview/javascript/src/chapter-one/one-point-five.js
// Implement a method to perform basic string compression using the counts of
// repeated characters. rg. aabcccccaaa will become a2b1c5a3
// if the compressed string isn't less ni length than the original, return the
// original string

export const stringCompressor = (string) => {
  const characters = string.split("");

  const characterMap = characters.reduce((charMap, character) => {
    return charMap[character]
      ? { ...charMap, [character]: charMap[character] + 1 }
      : { ...charMap, [character]: 1 };
  }, {});

  const result = Object.keys(characterMap).reduce(
    (compressedString, character) => {
      return `${compressedString}${character}${characterMap[character]}`;
    },
    ""
  );

  return result.length >= string.length ? string : result;
};

File: CrackTheCodingInterview/haskell/src/Chapter1/OnePointFive.hs
{-# LANGUAGE OverloadedStrings #-}

module Chapter1.OnePointFive
  ( stringCompressor
  )
where

--Implement a method to perform basic string compression of repeated characters
-- return the string if the the compressed string is larger than the original


import qualified Data.Text                     as T
import Debug.Trace
import Data.Char

stringCompressor' :: String -> String -> Char -> Int -> String -> String
stringCompressor' compressedString [] lastCharacter number originalString = if length originalString <= length updatedCompressedString then originalString else updatedCompressedString 
   where updatedCompressedString = lastCharacter:intToDigit number:compressedString
stringCompressor' compressedString (next:nextxs) previousChar number originalString 
 | previousChar == next = stringCompressor' compressedString nextxs next (number + 1) originalString
 | otherwise =  stringCompressor' updatedCompressedString nextxs next 1 originalString 
   where updatedCompressedString = previousChar:intToDigit number:compressedString

stringCompressor string = stringCompressor' "" (tail string)  first 1 string
  where first = head string



